### ISSUE TITLE
Possible race condition

### ISSUE THREAD

**Issue by juliandroid** [pdf/websocketrwc#1]
I might be wrong and/or miss something, but it looks to me that the mutex is unlocked too early.

_, r, err = c.ws.NextReader()
...
c.rmutex.Unlock()
...
_, err = io.Copy(c.buf, r)

If someone calls in parallel NextReader() before the io.Copy() runs then the old (unconsumed) frame will be lost. I'm right, not am I?

Edit: Also, to avoid posting another ticket, I see inside Read(), returning original value of "n" argument.
Since the Read() is supposed to be non-blocking and might receive and return, respectively, less than "n" bytes.

**Comment by pdf** [comment]
Hey, humble apologies for the delay getting back to you - apparently repos created with the `hub` command aren't watched by default, so I didn't get any notification for the issue.

Looks like you're correct regarding the mutex, should definitely guard around the copy, I'll push that shortly.

As for returning `n`, we're just returning the zero value.  This is correct, because we've not actually read anything into `p`, just pushed data into the buffer.

Looking at this again though, I probably need to wrap bytes.Buffer with a mutex to make it thread-safe since it's not by default, I'll sort that out too.

**Comment by juliandroid** [comment]
I did't have time to check the example, but I've noticed a lot of changes and a lot of lock/unlocks.
One thing that caught my attention was:

```
                c.rmutex.Lock()
		select {
		    case <-c.done:
			    err = ErrClosing
		    default:
			    err = c.ws.SetReadDeadline(time.Now().Add(PongTimeout))
			    if err == nil {
				    _, r, err = c.ws.NextReader()
			    }
		}
		if err != nil {
			return n, err
		}
		_, err = io.Copy(c.buf, r)
		c.rmutex.Unlock()

```
It seems that the c.rmutex might stay locked if err != nil
I guess a defer Unlock statement might be more appropriate.

**Comment by pdf** [comment]
I should not review code at 2am.  The code was already broken, I've just changed it to use defers like it should have in the first place.
